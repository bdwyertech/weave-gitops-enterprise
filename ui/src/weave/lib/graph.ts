// A collection of helper functions to render a graph of kubernetes objects
// with in the context of their parent-child relationships.
import _ from "lodash";
import { convertResponse } from "../hooks/objects";
import { Core } from "./api/core/core.pb";
import { GroupVersionKind, Kind } from "./api/core/types.pb";
import { FluxObject } from "./objects";

// Kubernetes does not allow us to query children by parents.
// We keep a list of common parent-child relationships
// to look up children recursively.
export const PARENT_CHILD_LOOKUP = {
  Deployment: {
    group: "apps",
    version: "v1",
    kind: "Deployment",
    children: [
      {
        group: "apps",
        version: "v1",
        kind: "ReplicaSet",
        children: [{ version: "v1", kind: "Pod" }],
      },
    ],
  },
  StatefulSet: {
    group: "apps",
    version: "v1",
    kind: "StatefulSet",
    children: [{ version: "v1", kind: "Pod" }],
  },
};

const objID = (obj: FluxObject) =>
  `${obj.clusterName}/${obj.namespace}/${obj.type}/${obj.name}`;

const sortFn = (a: FluxObject, b: FluxObject) => {
  return objID(a).localeCompare(objID(b));
};

export const getChildrenRecursive = async (
  client: typeof Core,
  namespace: string,

  object: FluxObject,
  clusterName: string,
  lookup: any
) => {
  const children: FluxObject[] = [];

  const k = lookup[object.type as string];

  if (k && k.children) {
    for (let i = 0; i < k.children.length; i++) {
      const child: GroupVersionKind = k.children[i];

      const res = await client.GetChildObjects({
        parentUid: object.uid,
        namespace,
        groupVersionKind: child,
        clusterName: clusterName,
      });

      if (res.objects) {
        for (let q = 0; q < res.objects.length; q++) {
          const c = convertResponse("", res.objects[q]);
          const childKind = child.kind as string;
          await getChildrenRecursive(client, namespace, c, clusterName, {
            [childKind]: child,
          });
          children.push(c);
        }
      }
    }
  }
  // Mutates
  children.sort(sortFn);
  object.children = children;
};

// Gets the "child" objects that result from an Application
export const getChildren = async (
  client: typeof Core,
  automationName: string,
  namespace: string,
  automationKind: Kind,
  kinds: GroupVersionKind[],
  clusterName: string
): Promise<FluxObject[]> => {
  const { objects } = await client.GetReconciledObjects({
    automationName,
    namespace,
    automationKind,
    kinds,
    clusterName,
  });

  const fluxObjs = _.map(objects, (o) => convertResponse("", o));
  fluxObjs.sort(sortFn);

  const result: FluxObject[] = [];
  for (let o = 0; o < fluxObjs.length; o++) {
    const obj = fluxObjs[o];
    await getChildrenRecursive(
      client,
      namespace,
      obj,
      clusterName,
      PARENT_CHILD_LOOKUP
    );
    result.push(obj);
  }
  const flat = _.flatten(result);

  return flat;
};
