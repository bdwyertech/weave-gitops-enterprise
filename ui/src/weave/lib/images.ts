/* eslint-disable */
// @ts-nocheck

import signInWheel from "data-url:../../../images/SignInWheel.svg";
import bg from "data-url:../../../images/bg-circles.png";
import bgDark from "data-url:../../../images/bg-dark.png";
import darkModeIcon from "data-url:../../../images/dark-mode-icon.svg";
import failedSrc from "data-url:../../../images/failed.svg";
import lightModeIcon from "data-url:../../../images/light-mode-icon.svg";
import logoDark from "data-url:../../../images/logoDark.svg";
import logoLight from "data-url:../../../images/logoLight.svg";
import logotype from "data-url:../../../images/logotype.svg";
import logotypeLight from "data-url:../../../images/logotypeLight.svg";
import pendingAction from "data-url:../../../images/pending-action.svg";
import reconcileSrc from "data-url:../../../images/reconcile.svg";
import signInWheelDark from "data-url:../../../images/signInWheelDark.svg";
import successSrc from "data-url:../../../images/success.svg";
import suspendedSrc from "data-url:../../../images/suspended.svg";
export default {
  bg,
  bgDark,
  failedSrc,
  logoLight,
  logoDark,
  logotype,
  logotypeLight,
  signInWheel,
  signInWheelDark,
  successSrc,
  suspendedSrc,
  reconcileSrc,
  pendingAction,
  darkModeIcon,
  lightModeIcon,
};
