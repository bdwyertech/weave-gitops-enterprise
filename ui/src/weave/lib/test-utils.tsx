import { ThemeProvider as MuiThemeProvider } from "@mui/material";
import { createMemoryHistory } from "history";
import _ from "lodash";
import * as React from "react";
import { QueryClient, QueryClientProvider } from "react-query";
import { Router } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import AppContextProvider, {
  AppProps,
  ThemeTypes,
} from "../contexts/AppContext";
import { CoreClientContext } from "../contexts/CoreClientContext";
import {
  Core,
  GetChildObjectsRequest,
  GetChildObjectsResponse,
  GetReconciledObjectsRequest,
  GetReconciledObjectsResponse,
  GetVersionRequest,
  GetVersionResponse,
  ListObjectsRequest,
  ListObjectsResponse,
} from "./api/core/core.pb";
import theme, { muiTheme } from "./theme";
import { RequestError } from "./types";
import { jsx } from "@emotion/react";
export type CoreOverrides = {
  GetChildObjects?: (req: GetChildObjectsRequest) => GetChildObjectsResponse;
  GetReconciledObjects?: (
    req: GetReconciledObjectsRequest
  ) => GetReconciledObjectsResponse;
  GetVersion?: (req: GetVersionRequest) => GetVersionResponse;
  ListObjects?: (req: ListObjectsRequest) => ListObjectsResponse;
};

export const createCoreMockClient = (
  ovr: CoreOverrides,
  error?: RequestError
): typeof Core => {
  const promisified = _.reduce(
    ovr,
    (result: { [key: string]: any }, handlerFn: any, method: string) => {
      result[method as string] = (req: GetChildObjectsRequest & GetReconciledObjectsRequest & ListObjectsRequest) => {
        if (error) {
          return new Promise((_, reject) => reject(error));
        }
        return new Promise((accept) => {
          if (handlerFn) {
            accept(handlerFn(req) as any);
          }
        });
      };

      return result;
    },
    {}
  );

  return promisified as typeof Core;
};

export function withTheme(element: any, mode: ThemeTypes = ThemeTypes.Light) {
  const appliedTheme = theme(mode);
  return (
    <ThemeProvider theme={appliedTheme}>
      <MuiThemeProvider theme={muiTheme(appliedTheme.colors, mode)}>
        {element}
      </MuiThemeProvider>
    </ThemeProvider>
  );
}

type TestContextProps = AppProps & {
  api?: typeof Core;
  featureFlags?: { [key: string]: string };
};

export function withContext(
  TestComponent: any,
  url: string,
  { api, featureFlags, ...appProps }: TestContextProps
) {
  const history = createMemoryHistory({ initialEntries: [url] });
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });
  const isElement = React.isValidElement(TestComponent);
  window.matchMedia = jest.fn();
  //@ts-ignore
  window.matchMedia.mockReturnValue({ matches: false });
  return (
    <Router location={url} navigator={history}>
      <AppContextProvider footer={<></>} {...appProps}>
        <QueryClientProvider client={queryClient}>
          <CoreClientContext.Provider
            value={{ api: api!, featureFlags: featureFlags || {} }}
          >
            {isElement ? TestComponent : <TestComponent />}
          </CoreClientContext.Provider>
        </QueryClientProvider>
      </AppContextProvider>
    </Router>
  );
}
