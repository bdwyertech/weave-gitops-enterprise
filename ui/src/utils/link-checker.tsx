//import { isAllowedLink } from '../gitops.d';

import { isAllowedLink } from "../weave/lib/utils";

export const openLinkHandler = (url: string) => {
  if (!isAllowedLink(url)) {
    return () => {
      return;
    };
  }
  return () => window.open(url, '_blank', 'noopener,noreferrer');
};
