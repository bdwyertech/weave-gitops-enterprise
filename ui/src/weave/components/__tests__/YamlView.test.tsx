import "jest-styled-components";
import React from "react";
import renderer from "react-test-renderer";
import { Kind } from "../../lib/api/core/types.pb";
import { withContext, withTheme } from "../../lib/test-utils";
import { createYamlCommand } from "../../lib/utils";
import YamlView from "../YamlView";

describe("YamlView", () => {
  describe("snapshots", () => {
    it("renders", () => {
      const tree = renderer
        .create(
          withTheme(
            withContext(
              <YamlView
                header={createYamlCommand(
                  Kind.Kustomization,
                  "podinfo",
                  "flux-system"
                )}
                yaml="yaml\nyaml\nyaml\n"
              />,
              "",
              {}
            )
          )
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
