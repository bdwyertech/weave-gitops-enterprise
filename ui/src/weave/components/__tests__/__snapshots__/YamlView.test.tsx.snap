// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YamlView snapshots renders 1`] = `
.c1 {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 4px;
  width: 100%;
}

.c3 {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.c4 svg {
  height: 12px;
  width: 12px;
}

.c4 svg path.path-fill,
.c4 svg line.path-fill,
.c4 svg polygon.path-fill,
.c4 svg rect.path-fill,
.c4 svg circle.path-fill,
.c4 svg polyline.path-fill {
  fill: !important;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.c4 svg path.stroke-fill,
.c4 svg line.stroke-fill,
.c4 svg polygon.stroke-fill,
.c4 svg rect.stroke-fill,
.c4 svg circle.stroke-fill,
.c4 svg polyline.stroke-fill {
  stroke: !important;
  transition: stroke 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.c4 svg rect.rect-height {
  height: 12px;
  width: 12px;
}

.c4.downward {
  transform: rotate(180deg);
}

.c4.upward {
  transform: initial;
}

.c4 img {
  width: 12px;
}

.c5 svg {
  height: 16px;
  width: 16px;
}

.c5 svg path.path-fill,
.c5 svg line.path-fill,
.c5 svg polygon.path-fill,
.c5 svg rect.path-fill,
.c5 svg circle.path-fill,
.c5 svg polyline.path-fill {
  fill: !important;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.c5 svg path.stroke-fill,
.c5 svg line.stroke-fill,
.c5 svg polygon.stroke-fill,
.c5 svg rect.stroke-fill,
.c5 svg circle.stroke-fill,
.c5 svg polyline.stroke-fill {
  stroke: !important;
  transition: stroke 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.c5 svg rect.rect-height {
  height: 16px;
  width: 16px;
}

.c5.downward {
  transform: rotate(180deg);
}

.c5.upward {
  transform: initial;
}

.c5 img {
  width: 16px;
}

.c2 {
  background: #F6F7F9;
  padding: 12px;
  border-bottom: 1px solid #d8d8d8;
  font-family: monospace;
  color: #009CCC;
  text-overflow: ellipsis;
}

.c0 {
  margin-bottom: 12px;
  width: 100%;
  font-size: 12px;
  border: 1px solid #d8d8d8;
  border-radius: 8px;
  overflow: hidden;
}

.c0 .code-wrapper {
  position: relative;
}

.c0 .copy-wrapper {
  position: absolute;
  right: 4px;
  top: 8px;
  background: #F6F7F9;
  padding: 4px 8px;
  border-radius: 2px;
}

.c0 pre {
  padding: 24px 12px!important;
}

<div
  className="c0 UnstyledYamlView"
>
  <div
    className="c1 c2"
  >
    kubectl get kustomization podinfo -n flux-system -o yaml
    <div
      onClick={[Function]}
      style={
        {
          "cursor": "pointer",
        }
      }
    >
      <div
        className="c3 c4"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="FileCopyOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12zm-1 4H8c-1.1 0-1.99.9-1.99 2L6 21c0 1.1.89 2 1.99 2H19c1.1 0 2-.9 2-2V11zM8 21V7h6v5h5v9z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    className="code-wrapper"
  >
    <div
      className="copy-wrapper"
    >
      <div
        onClick={[Function]}
        style={
          {
            "cursor": "pointer",
          }
        }
      >
        <div
          className="c3 c5"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="FileCopyOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12zm-1 4H8c-1.1 0-1.99.9-1.99 2L6 21c0 1.1.89 2 1.99 2H19c1.1 0 2-.9 2-2V11zM8 21V7h6v5h5v9z"
            />
          </svg>
        </div>
      </div>
    </div>
    <pre
      style={
        {
          "background": "#fefefe",
          "backgroundColor": "transparent",
          "color": "#545454",
          "display": "block",
          "margin": 0,
          "overflowX": "auto",
          "padding": "0.5em",
        }
      }
    >
      <code
        style={
          {
            "whiteSpace": "pre",
            "wordBreak": "break-word",
          }
        }
      >
        <span
          style={
            {
              "textWrap": "wrap",
            }
          }
        >
          <span
            style={
              {
                "color": "#008000",
              }
            }
          >
            yaml\\nyaml\\nyaml\\n
          </span>
        </span>
      </code>
    </pre>
  </div>
</div>
`;
